kuzzleio/kuzzle
# Add ability to use errorsManager in plugins

## What does this PR do ?

PR on 2-dev https://github.com/kuzzleio/kuzzle/pull/1421

For now the new `errorsManager` was only exposed in the kuzzle core.
This PR allows users to use it inside their plugins by defining the custom errors in the `manifest.json` file.

For instance, if you want to use it in the `kuzzle-plugin-logger`, you should have an `errors` field in its `manifest.json`. The file should look:
```
{
    "name": "kuzzle-plugin-logger",
    "kuzzleVersion": ">=1.0.0 <2.0.0",
    "privileged": false,
    "errors": {
        "unknown_service": {
            "code": 1,
            "message": "[kuzzle-plugin-logger] Unable to find service %s",
            "class": "InternalError"
        },
        "some_error": {
            "code": 2,
            "message": "Some error occured %s",
            "class": "BadRequestError"
	},
        "some_other_error": {
            "code": 3,
            "message": "Some other error occured %s",
            "class": "ForbiddenError"
	}
    }
}

```
In the `plugins.json` (so the `plugins` domain) will be added a `kuzzle-plugin-logger` subdomain with its `errors` field filled by the one defined in the `manifest.json`.

The code of the subdomain is either `0x00` or defined in the `kuzzlerc`.
For instance, your kuzzlerc would be :

```
{
  "plugins": {
    "my-plugins": {
      "foo": "bar",
      "gordon": "freeman",
      "some": "stuff",
      "codes": {
          "my-plugins": 42
      }
    }
  }
}
```

For officials released plugins, they are now defined in the `default.config` of Kuzzle

### How should this be manually tested?

Guess what: in order to test it you need... a plugin. Yes.

For instance you can test it with the `kuzzle-plugin-logger`.
Go to the root of the plugin, and add an `errors` field in the `manifest.json` like above.

Now, if you want to throw an error, you will have to write

`context.errorsManager.throw(errorName, placeholders)`

[So here](https://github.com/kuzzleio/kuzzle-plugin-logger/blob/master/lib/index.js#L56), you could replace the line by 
`context.errorsManager.throw(unknown_service, serviceName)`

Trigger that error, and you will see the right message with the right error name
